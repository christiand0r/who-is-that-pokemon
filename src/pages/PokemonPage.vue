<template>
  <template v-if="!pokemon">
    <div class="loader">
      <div class="icon">
        <img src="@/assets/pokeball.gif" alt="Pokeball" />
      </div>
      <h1>Cargando...</h1>
    </div>
  </template>

  <template v-else>
    <PokemonPicture
      :pokemon-image="pokemon.sprite"
      :show-pokemon="showPokemon"
    />
    <div class="pokemon-container">
      <PokemonOptions
        :pokemons="pokemonArray"
        :pokemon-id="pokemon.id"
        :is-correct="isCorrect"
        @selected-pokemon="checkAnswer"
      />
      <div class="play-again">
        <button v-if="playAgain" @click="newGame">Jugar de nuevo</button>
      </div>
    </div>
  </template>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from "@/helpers/getPokemonsOptions";

export default {
  name: "Pokemon Page",
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemon: null,
      isCorrect: null,
      pokemonArray: [],
      playAgain: false,
      showPokemon: false,
    };
  },
  mounted() {
    this.mixPokemonArray();
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArray = await getPokemonOptions();

      const rndInt = Math.floor(Math.random() * 4);

      this.pokemon = this.pokemonArray[rndInt];
    },
    checkAnswer(selectedId) {
      this.playAgain = true;
      this.showPokemon = true;

      this.pokemon.id === selectedId
        ? (this.isCorrect = true)
        : (this.isCorrect = false);
    },
    newGame() {
      //Reset State
      this.pokemon = null;
      this.pokemonArray = [];
      this.showPokemon = false;
      this.isCorrect = null;
      this.playAgain = false;

      //Call sort method
      this.mixPokemonArray();
    },
  },
};
</script>

<style scoped>
.pokemon-container {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  padding: 2rem;
  z-index: 4;
  width: 50%;
}
.loader {
  display: flex;
  align-items: center;
  gap: 2rem;
  justify-content: center;
  min-height: 100vh;
  text-align: center;
  overflow: hidden;
}

.loader {
  color: #fff;
  flex-direction: column;
  gap: 0;
}

.loader h1 {
  margin: 0;
}

.icon {
  width: 80px;
  height: 80px;
}

.icon img {
  height: 100%;
  object-fit: cover;
}

.play-again {
  display: flex;
  justify-content: end;
  max-width: 480px;
  width: 100%;
}

.play-again button {
  background-color: var(--lapis);
  border-radius: var(--normal-radius);
  border: none;
  color: var(--ghost);
  font-size: 1.8rem;
  font-weight: var(--normal);
  padding: 1.5rem;
  text-transform: uppercase;
  transition: 0.3s background-color ease;
}

.play-again button:hover {
  background-color: var(--blue);
  cursor: pointer;
}

@media screen and (max-width: 760px) {
  .pokemon-container {
    width: 100%;
  }
}
</style>
