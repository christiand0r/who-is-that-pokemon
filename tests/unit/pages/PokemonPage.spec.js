import { shallowMount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage.vue";
import { mockPokemons } from "../mocks/mockPokemons.mock";

describe("Pokemon Page - Component", () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test("debe llamar a mixPokemonArray cuando sea montado", () => {
    //Spy the function mixPokemonArray
    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      "mixPokemonArray"
    );

    //Invoque again the wrapper cause' need check if when mounted the function is called
    const wrapper = shallowMount(PokemonPage);

    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test("debe de ser igual al snapshot cuando cargan los pokemones", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemon: mockPokemons[3],
          isCorrect: null,
          pokemonArray: mockPokemons,
          playAgain: false,
          showPokemon: false,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de mostrar los componentes PokemonPicture y PokemonOptions con sus respectivos atributos", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemon: mockPokemons[3],
          isCorrect: null,
          pokemonArray: mockPokemons,
          playAgain: false,
          showPokemon: false,
        };
      },
    });

    const pokemonPicture = wrapper.find("pokemon-picture-stub");
    const pokemonOptions = wrapper.find("pokemon-options-stub");

    //Pokemon Picture Component
    expect(pokemonPicture.exists()).toBe(true);

    //Pokemon Options Component
    expect(pokemonOptions.exists()).toBe(true);

    //Evaluate expect with vm
    expect(pokemonPicture.attributes("pokemonimage")).toBe(
      wrapper.vm.pokemon.sprite
    );

    expect(pokemonOptions.attributes("pokemons")).toBeTruthy();

    //Evaluate expect with mock
    expect(pokemonOptions.attributes("pokemonid")).toBe(
      String(mockPokemons[3].id)
    );
  });

  test("método CheckAnswer", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemon: mockPokemons[3],
          isCorrect: null,
          pokemonArray: mockPokemons,
          playAgain: false,
          showPokemon: false,
        };
      },
    });
    //Set the correct id then state isCorrect must be true
    await wrapper.vm.checkAnswer(mockPokemons[3].id);
    expect(wrapper.vm.isCorrect).toBe(true);

    //Set the incorrect id then state isCorrect must be false
    await wrapper.vm.checkAnswer(mockPokemons[0].id);
    expect(wrapper.vm.isCorrect).toBe(false);

    //If checkAnswer is called the button play must be exist
    expect(wrapper.vm.playAgain).toBe(true);
    expect(wrapper.find(".play-again button").exists()).toBe(true);

    //If checkAnswer is called the state showPokemon must be true
    expect(wrapper.vm.showPokemon).toBe(true);
  });

  test("método newGame", async () => {
    //Spy for method newGame
    const newGameSpy = jest.spyOn(PokemonPage.methods, "newGame");

    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemon: mockPokemons[3],
          isCorrect: null,
          pokemonArray: mockPokemons,
          playAgain: false,
          showPokemon: false,
        };
      },
    });

    await wrapper.vm.newGame();

    expect(newGameSpy).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.pokemon).not.toEqual(mockPokemons[3]);
  });
});
