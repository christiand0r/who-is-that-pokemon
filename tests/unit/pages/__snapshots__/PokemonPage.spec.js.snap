// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Page - Component debe de ser igual al snapshot cuando cargan los pokemones 1`] = `
<pokemon-picture-stub pokemonimage="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/25.svg" showpokemon="false"></pokemon-picture-stub>
<div class="pokemon-container">
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" pokemonid="25"></pokemon-options-stub>
  <div class="play-again">
    <!--v-if-->
  </div>
</div>
`;
